### PR TITLE
feat: Babel plugin

### LINKED ISSUE #251: Extra, wrapping `<div>` for untrusted components
This isn't a critical blocker or bug - more so something that could be a nuisance when running into more advanced styling use cases. When a component is rendered with `trust` enabled, the DOM renders as you'd expect:

Screen Shot 2024-02-05 at 3 13 17 PM

However, when a component doesn't have `trust` enabled, there's an extra `<div>` with no attributes that wraps your component's DOM: 

Screen Shot 2024-02-05 at 3 13 42 PM

Would it be possible to get rid of this extra, wrapping `<div>` that has no attributes?

### LINKED ISSUE #298: Components starting with two lines of comments do not render
First noted on the Phosphor icons demo

> https://bos-web-engine.vercel.app/bwe-demos.near/NPM.PhosphorIcons?showCode=true
> 
> No horse :(

After some testing, it turns out the issue isn't with the import there but with the comments at the top. Reproduced with lodash demo below

## Functional

Snippet of `srcdoc` without leading comments where render succeeds. Comments are stripped and imports are hoisted
```ts
/******** BEGIN BOS SOURCE ********/
/******** The root Component definition is inlined here as [function BWEComponent() {...}] ********/
import __BWEModule__lodash from &quot;lodash&quot;;

import __BWEModule__lodash_partition from &quot;lodash/partition&quot;;


    /************************* bwe-web.near/NPM.Lodash (root) *************************/
    const BWEComponent = (() => {
      const _ = __BWEModule__lodash;
const partition = __BWEModule__lodash_partition;
      
      function BWEComponent() {
  return __Preact.createElement(React.Fragment, null, JSON.stringify(partition([1, 2, 3, 4], n => n % 2)), JSON.stringify(_.defaults({
    a: 1
  }, {
    a: 3,
    b: 2
  })));
}
      return BWEComponent;
    })();
  
/******** END BOS SOURCE ********/
```

## Error

Snippet of `srcdoc` with leading comments where render fails. Only first comment is stripped and imports are injected into component function where they are not allowed
### Input
```tsx
// one
// two
import _ from 'lodash';
import partition from 'lodash/partition';

export default function () {
  return (
    <>
      {JSON.stringify(partition([1, 2, 3, 4], (n) => n % 2))}
      {JSON.stringify(_.defaults({ a: 1 }, { a: 3, b: 2 }))}
    </>
  );
}
```
### Output
```ts
/******** BEGIN BOS SOURCE ********/
/******** The root Component definition is inlined here as [function BWEComponent() {...}] ********/

    /************************* bwe-web.near/NPM.Lodash (root) *************************/
    const BWEComponent = (() => {
      
      
      // two
import _ from 'lodash';
import partition from 'lodash/partition';
function BWEComponent() {
  return __Preact.createElement(React.Fragment, null, JSON.stringify(partition([1, 2, 3, 4], n => n % 2)), JSON.stringify(_.defaults({
    a: 1
  }, {
    a: 3,
    b: 2
  })));
}
      return BWEComponent;
    })();
  
/******** END BOS SOURCE ********/
```

### LINKED ISSUE #341: Bug: Trusted child components with updated state replace the root
I finished updating the styles for our sandboxed/trusted demo and noticed some funky behavior: https://bwe.near.dev/webengine.near/StateAndTrust.Root

- Click on any of the shapes on the right (sandboxed), and things work as expected
- Click on any of the shapes on the left (trusted), and the root component appears to be replaced by the child component

I created a simpler reproduction of this bug here: https://bwe.near.dev/webengine.near/TrustStateBugParent

Parent:

```tsx
import s from './styles.module.css';
import ChildComponent from './TrustStateBugChild';

function ParentComponent() {
  return (
    <div className={s.wrapper}>
      <p>Parent. This will dissappear when you click the button.</p>
      <ChildComponent trust={{ mode: 'trusted-author' }} />
    </div>
  );
}

export default ParentComponent as BWEComponent;
```

Child: 

```tsx
import s from './styles.module.css';
import {useState}from 'react';

function ChildComponent() {
  const [count, setCount] = useState(0);

  const increase = () => setCount((value) => value + 1);

  return (
    <div className={s.wrapper}>
      <button type="button" onClick={increase}>Count: {count}</button>
    </div>
  );
}

export default ChildComponent as BWEComponent;
```

Remove the `trust` prop and everything works as expected. I also tried playing around with setting `id` values, but that didn't help.

### LINKED ISSUE #346: Babel plugin for parsing Components
The current implementation uses regex to parse JS modules, leaving gaps in the implementation through which rudimentary syntax can break the compiler output by producing invalid JS based on bad parser assumptions. Maintaining this is not tenable in the long term, as new functionality only makes the current parser more brittle.

One solution is to move the parser logic into a custom Babel plugin, moving from parsing text to parsing the AST. The responsibilities remain the same:

- Determining the set of trusted child Components
- Mapping imported Component identifiers to BOS source
- Parsing and mapping `props`

This change not only directly addresses behavior seen in issues like #298, but unblocks and/or simplifies several other critical issues:
- #261 
- #341 
- #301 
- #251

### PR DESCRIPTION
**Pull request by andy-haynes**
This PR replaces the text-based parsing logic in the `compiler` package to a Babel plugin. By moving to an AST-based approach, we can resolve the jank inherent in the current approach (fixes #298, fixes #341, fixes #251 ) and confidently move forward on changes to code rewrites (e.g. directly injecting `Component` references for sandboxed renders, re-mapping props).

There shouldn't be any discernible difference in processing time. From my simple benchmarking, first-time transpilation ranges from 40-60ms regardless of the implementation used.

Fixes #346

### CODE REVIEW

File: packages/compiler/src/import.ts [hunk 1]
```diff
@@ -19,160 +12,45 @@ const PLUGIN_MODULES = new Map<string, string>([
...
+export function parseModuleImport(
+  modulePath: string,
+  imports: ImportExpression[]
+): ModuleImport {
+  let moduleName = extractModuleName(modulePath);
+  const isRelative = !!modulePath?.match(/^\.?\.\/(\.\.\/)*[a-z_$][\w\/]*$/gi);
+
+  // TODO check against Component name
+  const isCssModule =
+    modulePath.startsWith('./') && modulePath.endsWith('.module.css');
```
Comment (mpeterdev) [inline packages/compiler/src/import.ts:33]: going to start doing `import s from './🌭.module.css';` until this changes 🙂

File: packages/compiler/src/transpile.ts [hunk 2]
```diff
@@ -1,20 +1,354 @@
...
+         *   TL; DR
+         *   this preserves ancestry:
+         *   import Y from './Y';
+         *   export default function X() { return <Y />; }
+         *
+         *   this doesn't:
+         *   import Y from './Y';
+         *   const renderY = () => <Y />;
+         *   export default function X() { return <>{renderY()}</> }
+         */
```
Comment (mpeterdev) [inline packages/compiler/src/transpile.ts:145]: does this merit a dedicated issue? if so, please create one

Comment (andy-haynes) [reply packages/compiler/src/transpile.ts:145]: This didn't actually break in the scenarios I expected it to, so it's difficult to articulate the potential impact. I think for now it's a curiosity, but if I come up with a real case I'll write it up.

File: packages/compiler/src/compiler.ts [hunk 3]
```diff
@@ -132,41 +145,41 @@ export class ComponentCompiler {
...
+      children,
+      exports,
+      imports,
+      source: transpiledComponentSource,
+    } = this.getTranspiledComponentSource({
       componentPath,
       componentSource,
+      isComponentPathTrusted: isComponentPathTrusted
+        ? isComponentPathTrusted
+        : () => false,
```
Comment (pavelisnear) [inline packages/compiler/src/compiler.ts:158]: A small one, just to save a line:
```suggestion
      isComponentPathTrusted: isComponentPathTrusted || () => false,
```

File: packages/compiler/src/transpile.ts [hunk 4]
```diff
@@ -1,20 +1,354 @@
...
+        }
+      },
+      ImportDeclaration(path: { node: ImportDeclaration; remove(): void }) {
+        const {
+          node: { source, specifiers },
+        } = path;
+
+        const importExpressions = specifiers.map((specifier) => {
+          if (t.isImportSpecifier(specifier)) {
+            // TODO differentiate/handle namespaced & side-effect imports
```
Comment (pavelisnear) [inline packages/compiler/src/transpile.ts:305]: Might be worth creating a backlog task for that // TODO.

**Review by pavelisnear — APPROVED** [review]
Good job 👍

File: packages/compiler/src/transpile.ts (hunk 4 above)
Comment (andy-haynes) [reply packages/compiler/src/transpile.ts]: Pushed an update 🙂

**Pull request merged**